actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,489 @@
+		return nil, errs.New("Database is nil")
+	}
+	space, err := act.Db.Spaces().Load(act.Ctx, spaceID)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading space: "+err.Error())
+	}
+	groups, err := act.Db.WorkItemTypeGroups().List(act.Ctx, space.SpaceTemplateID)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading work item type: "+err.Error())
+	}
+	return groups, nil
+}
+
+func (act ActionStateToMetaState) loadWorkItemTypeByID(id uuid.UUID) (*workitem.WorkItemType, error) {
+	if act.Ctx == nil {
+		return nil, errs.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errs.New("Database is nil")
+	}
+	wit, err := act.Db.WorkItemTypes().Load(act.Ctx, id)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading work item type: "+err.Error())
+	}
+	return wit, nil
+}
+
+func (act ActionStateToMetaState) loadWorkItemByID(id uuid.UUID) (*workitem.WorkItem, error) {
+	if act.Ctx == nil {
+		return nil, errs.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errs.New("Database is nil")
+	}
+	wi, err := act.Db.WorkItems().LoadByID(act.Ctx, id)
+	if err != nil {
+		return nil, errs.Wrap(err, "error loading work item: "+err.Error())
+	}
+	return wi, nil
+}
+
+func (act ActionStateToMetaState) storeWorkItem(workitem *workitem.WorkItem) (*workitem.WorkItem, error) {
+	if act.Ctx == nil {
+		return nil, errs.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errs.New("Database is nil")
+	}
+	if act.UserID == nil {
+		return nil, errs.New("UserID is nil")
+	}
+	err := application.Transactional(act.Db, func(appl application.Application) error {
+		var err error
+		workitem, err = appl.WorkItems().Save(act.Ctx, workitem.SpaceID, *workitem, *act.UserID)
+		if err != nil {
+			return errs.Wrap(err, "error updating work item")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return workitem, nil
+}
+
+func (act ActionStateToMetaState) getValueListFromFieldType(wit *workitem.WorkItemType, fieldName string) ([]interface{}, error) {
+	fieldType := wit.Fields[fieldName].Type
+	switch t := fieldType.(type) {
+	case workitem.EnumType:
+		return t.Values, nil
+	}
+	return nil, errs.New("given field on workitemtype " + wit.ID.String() + " is not an enum field: " + fieldName)
+}
+
+func (act ActionStateToMetaState) getStateToMetastateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	wit, err := act.loadWorkItemTypeByID(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	stateList, err := act.getValueListFromFieldType(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromFieldType(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
```

Comment:
@michaelkleinhenz There is a chance of `index out of bounds` error here. We aren't checking the length of metastate array. We're assuming it will always be less than or equal to the states array.

Reply:
The above comment is probably incorrect.

Reply:
Why do you think it is incorrect?

Reply:
Fixed.